demo-circuit/src/naive_threshold_sig_w_key_rotation/tests.rs

```diff
@@ -0,0 +1,1171 @@
+use algebra::{Field, ToConstraintField, UniformRand};
+use cctp_primitives::{
+    proving_system::{
+        compute_proof_vk_size,
+        init::{get_g1_committer_key, load_g1_committer_key},
+        ProvingSystem,
+    },
+    utils::{commitment_tree::hash_vec, get_cert_data_hash_from_bt_root_and_custom_fields_hash},
+};
+use primitives::{
+    schnorr::field_based_schnorr::FieldBasedSchnorrSignature, FieldBasedHash,
+    FieldBasedSignatureScheme,
+};
+use r1cs_core::debug_circuit;
+use rand::{rngs::OsRng, thread_rng};
+use serial_test::serial;
+
+use crate::{
+    create_root_of_validator_keys, type_mapping::*, NaiveThresholdSignatureWKeyRotation,
+    ValidatorKeysSignatures, WithdrawalCertificateData, NULL_CONST,
+};
+use rand::Rng;
+
+use crate::{MAX_SEGMENT_SIZE, SUPPORTED_SEGMENT_SIZE};
+pub const VALIDATORS_SIZE: usize = 6;
+
+#[derive(Copy, Clone, Debug)]
+enum TestDataType {
+    AllRandomFirstCert,
+    AllRandomNextCert,
+    ValidSigEqThresholdNextCert,
+    ValidSigGreaterThresholdNextCert,
+    ValidSigSmallerThresholdNextCert,
+    ValidSigAndThresholdEqZeroNextCert,
+    ValidSigAndThresholdEqValidatorsNextCert,
+    ValidSigEqValidatorsThresholdEqZeroNextCert,
+    ValidSigEqZeroThresholdEqValidatorsNextCert,
+    ChangeSigningKeysNextCert,
+    ChangeMasterKeysNextCert,
+    ChangeSigningAndMasterKeysNextCert,
+}
+
+#[derive(Copy, Clone, Debug)]
+enum KeyToChange {
+    SignigKey,
+    MasterKey,
+    SignigAndMasterKeys,
+}
+
+fn generate_circuit_test_data(
+    test_data_type: TestDataType,
+) -> (usize, u32, usize, usize, bool, bool, bool, bool) {
+    let rng = &mut OsRng;
+    let max_pks = VALIDATORS_SIZE;
+    let num_custom_fields = 1;
+    let valid_sigs;
+    let threshold;
+    let is_first_certificate: bool;
+    let mut change_signing_keys = false;
+    let mut change_master_keys = false;
+
+    match test_data_type {
```

Comment:
Would be nice to randomize the `is_first_certificate` flag, or duplicate the test cases and test true/false both every time